libraries/libfc/include/fc/mutex.hpp

```diff
@@ -0,0 +1,227 @@
+#ifndef THREAD_SAFETY_ANALYSIS_MUTEX_HPP
+#define THREAD_SAFETY_ANALYSIS_MUTEX_HPP
```

Comment:
This is only if we use the clang `libc++`. Typically we build with gnu's `libstdc++`.

Reply:
Then we don't need it. As long as we have one build that checks this, then we don't need these wrappers.

Reply:
I think it would be worth a dedicated build to avoid having to provide these wrappers.

Reply:
I'm removing my hard objection to this. I would really like not to have the wrappers but can see the benefit of getting editor support. Another possibility is to move our default build to use `clang/libc++`. At this point, I'm ok with leaving these wrappers, but would like other opinions. @spoonincode ?

Reply:
Also in [mutex](https://github.com/llvm/llvm-project/blob/main/libcxx/include/mutex) only `scoped_lock` has the annotations.

Reply:
Then it may be worth reflecting on why the feature has been described as "mature enough to be deployed in an industrial setting" _for over 6 years_, yet LLVM has neglected to integrate it fully in to their own stdlib in that time.

Reply:
I think the expectation is that people will write their own wrappers. It is necessary anyways if you use clang with the default c++ stdlib from the system (could be gnu on ubuntu, Microsoft's on windows), so providing your own wrappers guarantees that you benefit from the Thread Safety Analysis when building with clang, regardless of which stdlib you use. 

in the [doc](https://clang.llvm.org/docs/ThreadSafetyAnalysis.html), they write:

> Thread safety analysis can be used with any threading library, but it does require that the threading API be wrapped in classes and methods which have the appropriate annotations. The following code provides mutex.h as an example; these methods should be filled in to call the appropriate underlying implementation.